writers/batchwriter/batchwriter.go

```diff
@@ -122,39 +122,46 @@ func (w *BatchWriter) Close(context.Context) error {
 }
 
 func (w *BatchWriter) worker(ctx context.Context, tableName string, ch <-chan *message.WriteInsert, flush <-chan chan bool) {
-	sizeBytes := int64(0)
-	resources := make([]*message.WriteInsert, 0, w.batchSize)
+	bytes, rows := writers.NewCapped(w.batchSizeBytes), writers.NewCapped(w.batchSize)
+	resources := make([]*message.WriteInsert, 0, w.batchSize) // at least we have 1 row per record
 	ticker := writers.NewTicker(w.batchTimeout)
 	defer ticker.Stop()
+
+	send := func() {
+		w.flushTable(ctx, tableName, resources)
+		clear(resources)
+		resources = resources[:0]
+		bytes.Reset()
+		rows.Reset()
+	}
 	for {
 		select {
 		case r, ok := <-ch:
 			if !ok {
-				if len(resources) > 0 {
+				if rows.Current() > 0 {
 					w.flushTable(ctx, tableName, resources)
 				}
 				return
 			}
 
-			if (w.batchSize > 0 && len(resources) >= w.batchSize) || (w.batchSizeBytes > 0 && sizeBytes+util.TotalRecordSize(r.Record) >= int64(w.batchSizeBytes)) {
-				w.flushTable(ctx, tableName, resources)
+			// we append data first not to call calculations twice
+			resources = append(resources, r)
```

Comment:
Doesn't this changes the logic to flush after the limit is reached, instead of before? i.e. we'll always exceed the limit before flushing? I think we want to flush before hitting it, no?

For BigQuery I think this means the write will fail (@hermanschaaf can confirm)

Reply:
Initially `writers.Cap` included `OverflownBy(a A)` method that checked this, I can add it back

Reply:
9dbac8d613ad90520e9c2daf08314d3528c00a72